### ISSUE TITLE
[trainer] algorithm.use_pf_ppo silently mis-pairs advantages with trajectories on the v1 trainer

### ISSUE THREAD

**Issue by YeonwooSung** [verl-project/verl#7956]
### System Info

Reproduced by reading and by executing the advantage code path on CPU (no GPU needed) at `23a341e3`.

```
verl: main @ 23a341e3
python: 3.12, torch 2.14 (CPU)
platform: macOS (arm64); the defect is device-independent — it is in driver-side advantage bookkeeping
```

### Information

- [x] The official example scripts
- [ ] My own modified scripts

### Tasks

- [x] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

**Summary:** with `algorithm.use_pf_ppo=true` and `algorithm.adv_estimator=gae` on the **v1** trainer (the default, `trainer.use_v1: true` — `verl/trainer/config/ppo_trainer.yaml:232`), each trajectory is trained against **another trajectory's advantage**. It fails silently: no error, no warning, and the metrics look plausible.

#### Why it happens

PF-PPO reweighting is implemented as a **resample of the whole batch with replacement**:

https://github.com/verl-project/verl/blob/23a341e3/verl/trainer/ppo/core_algos.py#L2303-L2305

```python
sample_indices = torch.multinomial(weights, batch_size, replacement=True)
resampled_batch = {key: tensor[sample_indices] for key, tensor in data.batch.items()}
```

The v0 trainer honours that contract: it resamples the entire `DataProto` and carries **that same object** forward into the critic/actor update (`verl/trainer/ppo/ray_trainer.py:229-233`, result flows on from `:1667`). Every field of row `i` comes from the same source row, so v0 is correct and is not affected by this report.

The v1 trainer cannot honour that contract, because its advantage step is **key-addressed**. In `verl/trainer/ppo/v1/trainer_base.py::_compute_advantage`:

- `:1761` — `response_mask` is captured from the TransferQueue read, i.e. in the **pre-resampling** row order.
- `:1790-1799` — `compute_advantage_for_multi_trajectories(..., config=self.config.algorithm)`.
- `:1810-1812` — `response_to_nested(data.batch[field], response_mask)` pairs the **resampled** advantages with the **stale** per-row response lengths. `verl/workers/utils/padding.py:223-227` slices row `i` to `response_lens[i]` with no length check, so this is silent.
- `:1814` — `tq.kv_batch_put(keys=batch.keys, ...)` writes those rows back under the **original** keys.

Only `advantages`/`returns` (plus `token_level_rewards` when `use_kl_in_reward`) are written back. The responses, `old_log_probs` and `ref_log_prob` stay in the TransferQueue under the original keys, unresampled. So the actor update pairs trajectory `k`'s tokens with some other trajectory's advantage.

The reason this is easy to miss: `grep -rn "use_pf_ppo" verl/trainer/ppo/v1/` returns **nothing**. v1 reaches the PF-PPO branch indirectly — `verl/trainer/ppo/v1/utils.py:22` imports v0's `compute_advantage`, and `:167-175` delegates every non-GRPO estimator to it, PF-PPO branch included.

#### Reproduction (CPU, no cluster)

```python
import numpy as np, torch
from unittest.mock import patch
from omegaconf import OmegaConf
from tensordict import TensorDict
from verl.protocol import DataProto
from verl.trainer.ppo import core_algos
from verl.trainer.ppo.v1.utils import compute_advantage_for_multi_trajectories

B, T = 8, 4
scores = torch.arange(B, dtype=torch.float32).unsqueeze(-1).repeat(1, T)
batch = TensorDict({"token_level_scores": scores.clone(), "token_level_rewards": scores.clone(),
                    "values": torch.zeros(B, T), "response_mask": torch.ones(B, T)}, batch_size=[B])
data = DataProto(batch=batch, non_tensor_batch={"uid": np.array([f"u{i}" for i in range(B)], dtype=object)})
cfg = OmegaConf.create({"use_pf_ppo": True, "pf_ppo": {"reweight_method": "pow", "weight_pow": 2.0}})

calls = []
real = core_algos.compute_pf_ppo_reweight_data
with patch.object(core_algos, "compute_pf_ppo_reweight_data",
                  side_effect=lambda *a, **k: (calls.append(1), real(*a, **k))[1]):
    out = compute_advantage_for_multi_trajectories(
        data, batch_keys=[f"u{i}_0_0" for i in range(B)],
        adv_estimator=core_algos.AdvantageEstimator.GAE, config=cfg)

print("pf_ppo applied on the v1 path:", len(calls))
print("batch_keys order :", [f"u{i}" for i in range(B)])
print("returned row -> source uid:", list(out.non_tensor_batch["uid"]))
```

Output:

```
pf_ppo applied on the v1 path: 1
batch_keys order : ['u0', 'u1', 'u2', 'u3', 'u4', 'u5', 'u6', 'u7']
returned row -> source uid: ['u7', 'u5', 'u7', 'u7', 'u7', 'u7', 'u7', 'u6']
```

Row 0 of the returned advantages belongs to `u7`, but `_compute_advantage` writes it back under key `u0_...`.

**Please note what is and is not executed here.** I proved by execution that the PF-PPO reweight *is applied on the v1 path* and that it *returns rows in a different order than `batch_keys`*. The resulting mis-pairing in the TransferQueue I established by **reading** the write-back at `trainer_base.py:1809-1814` — I did not run a full v1 training step.

#### Blast radius

- Affected: v1 trainer + `adv_estimator=gae` + `use_pf_ppo=true`.
- Not affected: v0 (`verl/trainer/ppo/ray_trainer.py`), and the experimental trainers, which call `compute_advantage` directly on a whole batch they carry forward (`verl/experimental/separation/ray_trainer.py:621`, inherited by `one_step_off_policy` and `fully_async_policy`).
- Not affected: non-GAE estimators. The PF-PPO branch sits inside the GAE arm of `compute_advantage`, so GRPO etc. ignore `use_pf_ppo` on v0 and v1 alike.

The key is public and documented, which is how a user lands here: `verl/trainer/config/algorithm.py:662-663` (docstring `:634-635`), `verl/trainer/config/ppo_trainer.yaml:126,129`, `verl/trainer/config/_generated_ppo_trainer.yaml:877-878` (and the megatron/torchtitan/veomni generated YAMLs), and `docs/ascend_tutorial/zh/dev_guide/model_dev/parameter_and_metrics.md:324-326`.

`git log -S use_pf_ppo` shows PF-PPO added in #1719/#1753, then touched only by config refactors (#2147, #2621, #5848) and the TransferQueue/v1 work (#3649, #5454). I could not find any commit, comment or doc that states v1 intentionally excludes PF-PPO — it looks like the v1 write-back was simply never considered against PF-PPO's resampling. If that is wrong and this **is** intentional, please say so and I will close this.

### Expected behavior

`algorithm.use_pf_ppo=true` on v1 should either reweight correctly or refuse to run. It should not silently train on mismatched advantages.

Two directions, and I would like a maintainer preference before this is settled:

**(A) Reject the combination on v1 (fail loud).** Raise from the v1 advantage path when `use_pf_ppo=true` and `adv_estimator=gae`, with an error naming the supported path (v0, `trainer.use_v1=False`). Small, no behaviour change for anyone whose run is currently correct, and it converts silent corruption into an immediate, actionable error.

**(B) Wire PF-PPO into v1 properly.** Make the resampling visible to the write-back — resample the keys alongside the rows, or apply reweighting where the full batch is still addressable, so responses and log-probs travel with their advantages. Correct in principle, but it touches the TransferQueue key contract and is a design call I do not want to make unilaterally.

I have a PR ready for **(A)**, with a CPU regression test, and I am happy to do **(B)** instead if that is preferred.

---

*AI assistance was used in investigating and writing up this issue. I have verified every file:line reference and run the reproduction above myself.*
